File: Src/Lexer.ts
export class Lexer {
    private input: string;
    private lineNumber: number;
    private messages: string[];

    constructor(input: string) {
        this.input = input;
        this.lineNumber = 1;
        this.messages = [];
    }

    public analyze() {
        let openBlocks = 0;

        for (let i = 0; i < this.input.length; i++) {
            const char = this.input[i];

            if (char === '{') {
                lexerOutput.debugLexer(`opening block found at line ${this.lineNumber}`);
                openBlocks++;
            } else if (char === '}') {
                if (openBlocks > 0) {
                    lexerOutput.debugLexer(`closing block notation found at line ${this.lineNumber}`);
                    openBlocks--;
                } else {
                    lexerOutput.error('Unmatched closing block', this.lineNumber);
                }
            }

            if (char === '\n') {
                this.lineNumber++;
            }
        }

        if (openBlocks > 0) {
            lexerOutput.error(`missing closing end of block notation at line ${this.lineNumber}`, this.lineNumber);
        }
        this.logMessages();
    }
    logMessages() {
        this.messages.forEach((message) => {
            console.log(message);
        });
    }

    getMessages() {
        return this.messages;
    }

    getLineNumber() {
        return this.lineNumber;
    }
    
}

export const lexerOutput = new class{
    private messages: string[] = [];

    private log(messageType: string, details: string) {
        this.messages.push(`${messageType}- ${details}`);
        this.updateOutputTextArea();
    }

    debugLexer(message: string) {
        this.log('DEBUG_LEXER', message);
    }

    warning(warningType: string, lineNumber: number) {
        const message = `WARNING- ${warningType} found at line ${lineNumber}`;
        this.log('WARNING', message);
    }

    error(errorType: string, lineNumber: number) {
        const message = `ERROR- ${errorType} found at line ${lineNumber}`;
        this.log('ERROR', message);
    }

    private updateOutputTextArea() {
        const outputTextArea = document.getElementById('outputTextarea') as HTMLTextAreaElement;
        if (outputTextArea) {
            outputTextArea.value = this.messages.join('\n');
        }
    }

    getMessages() {
        return this.messages;
    }
}

File: Src/Lexer.java
package Src;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Lexer {
    public void analyze(String filePath) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            int lineNumber = 1;
            boolean openBlockFound = false;
            int programCount = 1;
    
            System.out.println("Program " + programCount + " Lexical Analysis");
    
            while ((line = reader.readLine()) != null) {
                if (line.contains("{")) {
                    LexerUtils.debugLexer("Open Block Notation", "{", lineNumber);
                    openBlockFound = true;
                }
                if (line.contains("}")) {
                    if (openBlockFound) {
                        LexerUtils.debugLexer("Closing Block Notation", "}", lineNumber);
                        openBlockFound = false;
                    } else {
                        LexerUtils.error("Missing opening end of block notation", lineNumber);
                    }
                }
                if (line.contains("$")) {
                    LexerUtils.debugLexer("End of Program Notation", "$", lineNumber);
                    if (reader.ready()) {  // Check if there is more content after the EOP symbol
                        programCount++;
                        System.out.println("Program " + programCount + " Lexical Analysis");
                    }
                }
                lineNumber++;
            }
    
            if (openBlockFound) {
                LexerUtils.error("Missing closing end of block notation", lineNumber);
            }
        }
    }    
}



